### PR TITLE
Lexer performance testing with new streams

### PR DESCRIPTION
**Pull request by parrt**
This PR is an attempt to measure the speed of the new Java-target unicode streams by @bhamiltoncx and lexing with them. The legacy streams, such as `ANTLRFileStream` are tested as a baseline. Each lexer test is run multiple times (1500 or 500) and the highest 20% times are dropped; in Java these are likely  garbage collection or compilation pauses. The numbers seem reasonably stable across multiple runs (i.e., variance is low around that average time).

```
Warming up Java compiler....
   load_legacy_java_ascii average time    53us over 3500 loads of 29038 symbols from Parser.java
    load_legacy_java_utf8 average time    42us over 3500 loads of 29038 symbols from Parser.java
    load_legacy_java_utf8 average time   121us over 3500 loads of 13379 symbols from udhr_hin.txt
            load_new_utf8 average time   193us over 3500 loads of 29038 symbols from Parser.java
            load_new_utf8 average time   199us over 3500 loads of 13379 symbols from udhr_hin.txt

    lex_legacy_java_ascii average time   399us over 2000 runs of 29038 symbols
    lex_legacy_java_ascii average time   918us over 2000 runs of 29038 symbols DFA cleared
     lex_legacy_java_utf8 average time   377us over 2000 runs of 29038 symbols
     lex_legacy_java_utf8 average time   925us over 2000 runs of 29038 symbols DFA cleared
        lex_new_java_utf8 average time   460us over 2000 runs of 29038 symbols
        lex_new_java_utf8 average time   989us over 2000 runs of 29038 symbols DFA cleared

 lex_legacy_grapheme_utf8 average time  6862us over  400 runs of  6614 symbols from udhr_kor.txt
 lex_legacy_grapheme_utf8 average time  7023us over  400 runs of  6614 symbols from udhr_kor.txt DFA cleared
 lex_legacy_grapheme_utf8 average time  6290us over  400 runs of 13379 symbols from udhr_hin.txt
 lex_legacy_grapheme_utf8 average time  6238us over  400 runs of 13379 symbols from udhr_hin.txt DFA cleared
    lex_new_grapheme_utf8 average time  6863us over  400 runs of  6614 symbols from udhr_kor.txt
    lex_new_grapheme_utf8 average time  7014us over  400 runs of  6614 symbols from udhr_kor.txt DFA cleared
    lex_new_grapheme_utf8 average time  6187us over  400 runs of 13379 symbols from udhr_hin.txt
    lex_new_grapheme_utf8 average time  6284us over  400 runs of 13379 symbols from udhr_hin.txt DFA cleared
    lex_new_grapheme_utf8 average time    99us over  400 runs of    85 symbols from emoji.txt
    lex_new_grapheme_utf8 average time   111us over  400 runs of    85 symbols from emoji.txt DFA cleared
```

### CODE REVIEW

**Comment by bhamiltoncx** [comment]
Nice! I'll see if I can get this running under YourKit. Probably some low-hanging fruit in the new streams.

**Comment by parrt** [comment]
Okidoki.  I should also squash all those commits into a single one, which I would do shortly

**Comment by parrt** [comment]
@antlr/antlr-targets This might be a useful benchmark for your targets. Java profiling is notoriously challenging given the garbage collection and compilation pauses that occur at runtime. Other targets may not have this issue.

File: runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		                  (int)avg,
+		                  n,
+		                  input.size(),
+		                  fileName,
+		                  clearLexerDFACache ? " DFA cleared" : "");
+	}
+
+	public void new_grapheme_utf8(String fileName, int n, boolean clearLexerDFACache) throws Exception {
+		URL sampleJavaFile = TimeLexerSpeed.class.getClassLoader().getResource(PerfDir+"/"+fileName);
+		CharStream input = CharStreams.fromPath(Paths.get(sampleJavaFile.getFile()));
```
Comment (bhamiltoncx) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java]: Shouldn't this be part of the time measurement? This is basically the most expensive part of the new codepath, right?

File: runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		                  currentMethodName,
+		                  (int)avg,
+		                  n,
+		                  input.size(),
+		                  clearLexerDFACache ? " DFA cleared" : "");
+	}
+
+	public void legacy_grapheme_utf8(String fileName, int n, boolean clearLexerDFACache) throws Exception {
+		URL sampleJavaFile = TimeLexerSpeed.class.getClassLoader().getResource(PerfDir+"/"+fileName);
+		CharStream input = new ANTLRFileStream(sampleJavaFile.getFile());
```
Comment (bhamiltoncx) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java]: Ditto, this should probably be included in the time spent.

File: runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java (hunk 1 above)
Comment (parrt) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java]: whoops. i also missed a UTF8 on legacy stuff

Comment (parrt) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java]: Actually, i meant these to test lexing time so it's ok. I'll separately test the stream load time.

Comment (bhamiltoncx) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/java/api/perf/TimeLexerSpeed.java]: @parrt: 👌🏽

**Comment by parrt** [comment]
they do seem pretty close but the new streams are consistently a tiny bit slower during lexing, which doesn't necessarily make sense as you say. hmm... could be an artifact of the way I'm testing

**Comment by bhamiltoncx** [comment]
Yeah, I see the small difference. If it's consistent, it's probably real! The `CodePointCharStream` methods basically delegate to `IntBuffer`, which seems fairly optimal:

http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/nio/Buffer.java#Buffer

Maybe the JIT is inlining the methods in `ANTLRInputStream` which makes them a teeny bit faster? Is it worth chasing this down?

**Comment by parrt** [comment]
I'm noticing that the order in which I test these seems to matter for the loads, despite trying to warm up the compiler. damn this stuff is fiddly. I don't think it's worth chasing this performance down until we see it on a real problem. If there is a tiny advantage people can use the legacy streams.

**Pull request merged**